code/harness/Tester.cpp

```diff
@@ -760,65 +761,121 @@ int Tester::log_snapshot_save(string log_file) {
   // TODO(ashmrtn): What happens if this fails?
   // TODO(ashmrtn): Change device_clone to be an mmap of the disk we need to get
   // stuff on.
-  char device_clone[device_size];
-  unsigned int bytes_read = 0;
-  do {
-    int res = read(cow_brd_fd, device_clone + bytes_read,
-        device_size - bytes_read);
-    if (res < 0) {
-      cerr << "error reading from raw device to log disk snapshot" << endl;
-      break;
-    }
-    bytes_read += res;
-  } while (bytes_read < device_size);
-  ofstream log(log_file, std::ofstream::trunc);
-  log.write(device_clone, device_size);
-  bool err = log.fail();
-  log.flush();
-  log.close();
-  if (err) {
+  // device_size happens to be the number of 1k blocks on cow_brd (from original
+  // brd behavior...), so convert it to a number of bytes.
+  unsigned int dev_bytes = device_size * 2 * 512;
+  unsigned int bytes_done = 0;
+  const unsigned int buf_size = 4096;
+  unsigned int buf[buf_size];
+  int log_fd =
+    open(log_file.c_str(), O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+  if (log_fd < 0) {
+    cerr << "error opening log file" << endl;
     return LOG_CLONE_ERR;
   }
+
+  int res = lseek(cow_brd_fd, 0, SEEK_SET);
+  if (res < 0) {
+    cerr << "error seeking to start of test device" << endl;
+  }
+  while (bytes_done < dev_bytes) {
+    // Read a block of data from the base disk image.
+    unsigned int bytes = 0;
+    unsigned int new_amount = (dev_bytes < bytes_done + buf_size)
+                            ? dev_bytes - bytes_done
+                            : buf_size;
+    do {
+      int res = read(cow_brd_fd, buf + bytes, new_amount - bytes);
+      if (res < 0) {
+        cerr << "error reading from raw device to log disk snapshot" << endl;
+        break;
+      }
+      bytes += res;
+    } while (bytes < new_amount);
+
+    // Write a block of data to the log file.
+    bytes = 0;
+    do {
+      int res = write(log_fd, buf + bytes, new_amount - bytes);
+      if (res < 0) {
+        cerr << "error reading from raw device to log disk snapshot" << endl;
+        break;
```

Comment:
You might want to return with an error here? Even if you break, you still add the new_amount to bytes_done and return success!